refactor(napi/oxlint): add more types

overlookmotel opened the pull request:
Add more types to `visitor.ts`.

This is quite tricky, because the type of `compiledVisitor` changes over time.

Before calling `finalizeCompiledVisitor`, it's:

```
VisitFn | VisitFn[] | { enter: VisitFn | VisitFn[] | null, exit: VisitFn | VisitFn[] | null } | null
```

but after calling `finalizeCompiledVisitor`, all the arrays have been removed, so it's:

```
VisitFn | { enter: VisitFn | null, exit: VisitFn | null } | null
```

And which of `VisitFn` or `{ enter: VisitFn | null, exit: VisitFn | null }` it is depends on the index into `compiledVisitor`.

I've done my best, but this is quite an imperfect solution to these vagaries.

overlookmotel commented on napi/oxlint2/src-js/visitor.ts:
```diff
@@ -239,8 +253,8 @@ export function addVisitorToCompiled(visitor: any): void {
         compiledVisitor[typeId] = isExit
-          ? [existing, visitFn]
-          : [visitFn, existing];
+          ? [existing as VisitFn, visitFn]
+          : [visitFn, existing as VisitFn];
```
This is rather annoying.

Ideally want some way of telling typechecker at the top of this block that "`existing` is definitely `LeafCompiledVisitorEntry` here". Then the rest of the types can be inferred from that.

https://github.com/oxc-project/oxc/blob/ba55e92ac2c5853a6c06647277d5c342eae2dc81/napi/oxlint2/src-js/visitor.ts#L238-L258

I guess this would do it:

```js
const existing = compiledVisitor[typeId];
if (typeId < LEAF_NODE_TYPES_COUNT) {
  const existingLeaf = existing as LeafCompiledVisitorEntry;
  // ... use `existingLeaf` in the rest of this block
}
```

But this introduces another variable purely to satisfy TS. At runtime, this extra var is pointless, and may have a (small) runtime cost.

Is there anyway to communicate the type constraint to TS without generating more runtime code?

draphikas replied:
@overlookmotel I knew you were going to talk about the runtime cost of this function.

Unfortunatly, the `typeId < LEAF_NODE_TYPES_COUNT` check is not related to `existing`directly in types context.
Typescript don't know that if `typeId < LEAF_NODE_TYPES_COUNT` is true, that means the node is a leaf node.
The type guard function I showed you is a hack to enforce `existing` type without using it and infer the rest.

I don't actually know how to do that without creating this intermediate function.

But I agree that it's better if we can avoid type assertions, t can introduce some unexpected errors.

draphikas replied:
Just a thought
If we move down `const existing = compiledVisitor[typeId];`, we can use type assertion without creating extra variable :

Insted of : 
```ts
const existingEntry = compiledVisitor[typeId];
if (typeId < LEAF_NODE_TYPES_COUNT) {
  // Leaf node - store just 1 function, not enter+exit pair
  const existing = existingEntry as CompilingLeafVisitorEntry;
} else {
  // Not leaf node - store enter+exit pair
  const existing = existingEntry as CompilingNonLeafVisitorEntry;
}
```
We can do : 
```ts
if (typeId < LEAF_NODE_TYPES_COUNT) {
  // Leaf node - store just 1 function, not enter+exit pair
  const existing = compiledVisitor[typeId] as CompilingLeafVisitorEntry;
} else {
  // Not leaf node - store enter+exit pair
  const existing = compiledVisitor[typeId] as CompilingNonLeafVisitorEntry;
}
```
It's just some extra bytes but no more runtime cost

overlookmotel commented on napi/oxlint2/src-js/visitor.ts:
```diff
@@ -239,8 +253,8 @@ export function addVisitorToCompiled(visitor: any): void {
-          ? [existing, visitFn]
-          : [visitFn, existing];
+          ? [existing as VisitFn, visitFn]
+          : [visitFn, existing as VisitFn];
```
I've found a solution to this, but it's not very pleasing, because it involves creating pointless intermediate variables, which are a (small) cost at runtime.

camc314 commented:
### Merge activity

* **Jul 29, 10:40 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
refactor(napi/oxlint): add more types (#12577)

Add more types to `visitor.ts`.

This is quite tricky, because the type of `compiledVisitor` changes over time.

Before calling `finalizeCompiledVisitor`, it's:

```
VisitFn | VisitFn[] | { enter: VisitFn | VisitFn[] | null, exit: VisitFn | VisitFn[] | null } | null
```

but after calling `finalizeCompiledVisitor`, all the arrays have been removed, so it's:

```
VisitFn | { enter: VisitFn | null, exit: VisitFn | null } | null
```

And which of `VisitFn` or `{ enter: VisitFn | null, exit: VisitFn | null }` it is depends on the index into `compiledVisitor`.

I've done my best, but this is quite an imperfect solution to these vagaries.
